pkg/cloud/openstack/services/userdata/kubeadm.go

```diff
@@ -0,0 +1,164 @@
+
+	caCertHash, err := certificates.GenerateCertificateHash(clusterProviderSpec.CAKeyPair.Cert)
+	if err != nil {
+		return "", err
+	}
+
+	if bootstrapToken != "" && len(cluster.Status.APIEndpoints) == 0 {
+		return "", fmt.Errorf("no APIEndpoint set yet, waiting until APIEndpoints is set")
+	}
+
+	if isControlPlane {
+		if bootstrapToken == "" {
+			klog.Info("Machine is the first control plane machine for the cluster")
+
+			if !clusterProviderSpec.CAKeyPair.HasCertAndKey() {
+				return "", fmt.Errorf("failed to create controlplane kubeadm config, missing CAKeyPair")
+			}
+
+			clusterConfigurationCopy := clusterProviderSpec.ClusterConfiguration.DeepCopy()
+			// Set default values. If they are not set, these two properties are added with an
+			// empty string and the CoreOS ignition postprocesser fails with "could not find expected key"
+			if clusterConfigurationCopy.CertificatesDir == "" {
+				clusterConfigurationCopy.CertificatesDir = "/etc/kubernetes/pki"
+			}
+			if string(clusterConfigurationCopy.DNS.Type) == "" {
+				clusterConfigurationCopy.DNS.Type = kubeadmv1beta1.CoreDNS
+			}
+			kubeadm.SetClusterConfigurationOptions(
+				clusterConfigurationCopy,
+				kubeadm.WithKubernetesVersion(machine.Spec.Versions.ControlPlane),
+				kubeadm.WithAPIServerCertificateSANs(localIPV4Lookup),
+				kubeadm.WithAPIServerExtraArgs(map[string]string{"cloud-provider": cloudProvider}),
+				kubeadm.WithAPIServerExtraArgs(map[string]string{"cloud-config": cloudProviderConfig}),
+				kubeadm.WithAPIServerExtraVolumes([]kubeadmv1beta1.HostPathMount{
+					{
+						Name:      "cloud",
+						HostPath:  "/etc/kubernetes/cloud.conf",
+						MountPath: "/etc/kubernetes/cloud.conf",
+					},
+				}),
+				kubeadm.WithControllerManagerExtraArgs(map[string]string{"cloud-provider": cloudProvider}),
+				kubeadm.WithControllerManagerExtraArgs(map[string]string{"cloud-config": cloudProviderConfig}),
+				kubeadm.WithControllerManagerExtraVolumes([]kubeadmv1beta1.HostPathMount{
+					{
+						Name:      "cloud",
+						HostPath:  "/etc/kubernetes/cloud.conf",
+						MountPath: "/etc/kubernetes/cloud.conf",
+					},
+				}),
+				kubeadm.WithClusterNetworkFromClusterNetworkingConfig(cluster.Spec.ClusterNetwork),
+				kubeadm.WithKubernetesVersion(machine.Spec.Versions.ControlPlane),
+			)
+			clusterConfigYAML, err := kubeadm.ConfigurationToYAML(clusterConfigurationCopy)
+			if err != nil {
+				return "", err
+			}
+
+			kubeadm.SetInitConfigurationOptions(
+				&machineProviderSpec.KubeadmConfiguration.Init,
+				kubeadm.WithNodeRegistrationOptions(
+					kubeadm.NewNodeRegistration(
+						kubeadm.WithTaints(machine.Spec.Taints),
+						kubeadm.WithKubeletExtraArgs(map[string]string{"cloud-provider": cloudProvider}),
+						kubeadm.WithKubeletExtraArgs(map[string]string{"cloud-config": cloudProviderConfig}),
```

Comment:
Where does `cloudProviderConfig` initially get set? Can the user augment this, meaning, can they add additional things to the cloud config file prior to cluster creation?

Reply:
Config file is still written via user data. So whatever the user defines in the user data is written to disk. The only thing which is no fixed is the location of the file. But I think that's okay.